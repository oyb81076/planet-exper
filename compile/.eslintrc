{
  "env": {
    "browser": true,
    "es6": true,
    "node": true
  },
  "extends": [
    "airbnb-typescript",
    // "plugin:react-hooks/recommended",
    "plugin:@typescript-eslint/recommended",
    "plugin:@typescript-eslint/recommended-requiring-type-checking"
  ],
  "parserOptions": {
    "project": "./tsconfig.json"
  },
  "rules": {
    "@typescript-eslint/no-use-before-define": 0,
    "no-param-reassign": 0,
    "no-return-assign": 0
  },
  "overrides": [
    {
      "files": [
        "**/__tests__/**/*.ts"
      ],
      "env": {
        "jest": true
      },
      "rules": {
        "no-console": "off"
      }
    }
  ]
}
