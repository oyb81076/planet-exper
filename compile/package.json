{
  "name": "compile",
  "version": "1.0.0",
  "main": "index.js",
  "license": "MIT",
  "scripts": {
    "gen": "ts-node scripts/gen",
    "test": "jest",
    "lint": "eslint --ext .ts src"
  },
  "dependencies": {
    "acorn": "^7.3.1",
    "dayjs": "^1.8.28",
    "escodegen": "^1.14.2",
    "json5": "^2.1.3",
    "lodash": "^4.17.21",
    "nanoid": "^3.1.9",
    "parse5": "^6.0.0",
    "react": "^16.13.1"
  },
  "devDependencies": {
    "@babel/core": "^7.10.2",
    "@babel/plugin-proposal-class-properties": "^7.10.1",
    "@babel/preset-env": "^7.10.2",
    "@babel/preset-typescript": "^7.10.1",
    "@types/escodegen": "^0.0.6",
    "@types/estree": "^0.0.44",
    "@types/jest": "^25.2.3",
    "@types/lodash": "^4.14.155",
    "@types/node": "^14.0.12",
    "@types/parse5": "^5.0.3",
    "@typescript-eslint/eslint-plugin": "^3.1.0",
    "@typescript-eslint/parser": "^3.0.0",
    "eslint": "^6.8.0",
    "eslint-config-airbnb-typescript": "^8.0.2",
    "eslint-plugin-import": "^2.20.1",
    "eslint-plugin-jsx-a11y": "^6.2.3",
    "eslint-plugin-react": "^7.19.0",
    "eslint-plugin-react-hooks": "^4.0.4",
    "jest": "^26.0.1",
    "ts-node": "^8.10.2",
    "typescript": "^3.9.5"
  }
}
