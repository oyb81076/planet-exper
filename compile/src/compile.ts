import { parse, DefaultTreeElement, parseFragment } from 'parse5';
import { IHTMLFullDocument, IHTMLFragDocument } from './ast';
import { parseRoot } from './parse';

interface IConfig {
  parseExpression?: boolean;
  files?: Record<string, string>;
}

export function compileFull(content: string, cfg: IConfig = {}): IHTMLFullDocument {
  const doc = parse(content) as DefaultTreeElement;
  const node = doc.childNodes.find((x): x is DefaultTreeElement => x.nodeName === 'html');
  if (!node) {
    throw new Error('未发现<html>节点');
  }
  const {
    scripts, links, jsGlobals, element,
  } = parseRoot(node, cfg);
  return {
    type: '#html-document', scripts, links, jsGlobals, element,
  };
}
export function compileFragment(content: string, cfg: IConfig = {}): IHTMLFragDocument {
  const doc = parseFragment(content) as DefaultTreeElement;
  const node = doc.childNodes.find((x): x is DefaultTreeElement => x.nodeName[0] !== '#');
  if (!node) {
    throw new Error('未发现根节点');
  }
  const {
    scripts, links, jsGlobals, element, args,
  } = parseRoot(node, cfg);
  return {
    type: '#html-fragment', scripts, links, args, jsGlobals, element,
  };
}
