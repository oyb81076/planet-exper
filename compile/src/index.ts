export { compile } from './compile';
export { default as serialize } from './serialize';
