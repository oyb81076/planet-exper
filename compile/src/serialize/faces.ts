export interface ISerializeConfig {
  compress?: boolean;
  compactJS?: boolean;
  compactJSON?: boolean;
  quote?: boolean;
}

export interface ISerializer {
  compress: boolean;
  compactJS: boolean;
  compactJSON: boolean;
  quote: boolean;
}
