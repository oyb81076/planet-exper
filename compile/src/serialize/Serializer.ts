import {
  ILink,
  IScript,
  IJsGlobals,
  IElementAttribute,
  IDirectives,
  IDirectiveDisabled,
  IDirectiveEach,
  ITypeArg,
  IDirectiveIf,
  IDirectiveBind,
  IDirectiveSrc,
  IDirectiveHref,
  IDirectiveClasses,
  IElementOn,
  IDirectiveJavascript,
  IDirectiveText,
} from '../ast';
import recordExprToString from '../utils/recordExprToString';
import escapeAttrValue from './utils/escapeAttrValue';
import isSelfClosing from './utils/isSelfClosing';
import { srzEachValue } from '../directives/each';
import { srzIf } from '../directives/if';
import { srzBind } from '../directives/bind';
import { srzSrc } from '../directives/src';
import { srzHref } from '../directives/href';
import { srzClasses } from '../directives/classes';
import { srzOn } from '../directives/on';
import { srzJavascript } from '../directives/javascript';
import { srzText } from '../directives/text';
import { srzLinks } from '../directives/links';
import { ISerializeConfig, ISerializer } from './faces';
import { srzScripts } from '../directives/scripts';
import { srzArgs } from '../directives/args';
// html,head,body都不占用缩进
interface ITag {
  name: string;
  childRendered: boolean;
  inline: boolean;
}

export default class Serializer implements ISerializer {
  public readonly compress: boolean;

  public readonly compactJS: boolean;

  public readonly compactJSON: boolean;

  public readonly quote: boolean;

  private tags: Array<ITag> = [];

  private currentTag?: ITag;

  private indent = '';

  public html = '';

  constructor({
    compress = true, quote = !compress, compactJS = compress, compactJSON = compress,
  }: ISerializeConfig) {
    this.compress = compress;
    this.quote = quote;
    this.compactJS = compactJS;
    this.compactJSON = compactJSON;
  }

  doctype(): void {
    this.html += '<!DOCTYPE html>';
  }

  text(text: string): void {
    const { currentTag } = this;
    if (currentTag && !currentTag.childRendered) {
      currentTag.childRendered = true;
      currentTag.inline = false;
      this.html += '>';
    }
    this.newline();
    this.newIndent();
    this.html += text;
  }

  // 但行文本 <div>xxxxx</div>
  inlineText(text: string): void {
    const { currentTag } = this;
    if (text.length > 50 || !currentTag || currentTag.childRendered) {
      this.text(text);
    } else {
      currentTag.childRendered = true;
      this.html += `>${text}`;
    }
  }

  openTag(tag: string): void {
    const { currentTag } = this;
    if (currentTag && !currentTag.childRendered) {
      currentTag.childRendered = true;
      currentTag.inline = false;
      this.html += '>';
    }
    this.tags.push(this.currentTag = { name: tag, childRendered: false, inline: true });
    if (this.html) {
      this.newline();
      if (!this.compress && isIdentTag(tag)) {
        this.indent += '  ';
      }
      this.newIndent();
    }
    this.html += `<${tag}`;
  }

  closeTag(): void {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const { name, childRendered, inline } = this.currentTag!;
    this.tags.pop();
    this.currentTag = this.tags[this.tags.length - 1];
    if (!childRendered && isSelfClosing(name)) {
      this.html += '>';
    } else {
      if (childRendered) {
        if (!inline) {
          this.newline();
          this.newIndent();
        }
      } else {
        this.html += '>';
      }
      this.html += `</${name}>`;
    }
    if (!this.compress && isIdentTag(name)) {
      this.indent = this.indent.substr(2);
    }
  }

  private writeAttrName(name: string): void {
    this.html += ' ';
    this.html += name;
  }

  private writeAttrValue(value?: string): void {
    if (value) {
      this.html += '=';
      this.html += escapeAttrValue(value, this.quote);
    }
  }

  attrs(attrs: IElementAttribute[]): void {
    attrs.forEach(({ name, value }) => {
      this.writeAttr(name, value);
    });
  }

  directives(directives: IDirectives): void {
    this.xDisabled(directives.disabled);
    this.xEach(directives.each);
    this.xIf(directives.if);
    directives.binds.forEach((x) => this.xBind(x));
    this.xSrc(directives.src);
    this.xHref(directives.href);
    this.xClasses(directives.classes);
    directives.on.forEach((x) => this.xOn(x));
    this.xJavascript(directives.javascript);
    this.xText(directives.text);
  }

  private xDisabled(value?: IDirectiveDisabled): void {
    if (value) {
      this.writeAttrName('x-disabled');
    }
  }

  private xEach(dir?: IDirectiveEach): void {
    if (dir) {
      this.writeDirective('x-each', srzEachValue(dir, this));
    }
  }

  private xIf(dir?: IDirectiveIf): void {
    if (dir) {
      this.writeDirective('x-if', srzIf(dir, this));
    }
  }

  private xSrc(dir?: IDirectiveSrc): void {
    if (dir) {
      this.writeDirective('x-src', srzSrc(dir));
    }
  }

  private xHref(dir?: IDirectiveHref): void {
    if (dir) {
      this.writeDirective('x-href', srzHref(dir, this));
    }
  }

  private xClasses(value?: IDirectiveClasses): void {
    if (value) {
      this.writeDirective('x-classes', srzClasses(value, this));
    }
  }

  private xText(dir?: IDirectiveText): void {
    if (dir) {
      this.writeDirective('x-text', srzText(dir, this));
    }
  }

  private xBind(bind: IDirectiveBind): void {
    const [name, value] = srzBind(bind, this);
    this.writeDirective(name, value);
  }

  private xOn(dir: IElementOn): void {
    const [name, value] = srzOn(dir);
    this.writeDirective(name, value);
  }

  private xJavascript(dir?: IDirectiveJavascript): void {
    if (!dir) { return; }
    this.writeDirective('x-javascript', srzJavascript(dir));
  }

  writeDirective(name: string, value: string): void {
    this.writeAttrName(name);
    this.writeAttrValue(value);
  }

  writeAttr(name: string, value?: string): void {
    this.writeAttrName(name);
    this.writeAttrValue(value);
  }

  xKey(key: string): void {
    this.writeDirective('x-key', key);
  }

  xLinks(links: ILink[]): void {
    const value = srzLinks(links, this);
    if (value) {
      this.writeDirective('x-links', value);
    }
  }

  xScripts(scripts: IScript[]): void {
    const value = srzScripts(scripts, this);
    if (value) {
      this.writeDirective('x-scripts', value);
    }
  }

  xJsGlobals(globals: IJsGlobals): void {
    const out = globals.expr ? recordExprToString(globals.expr, this) : globals.content;
    if (out && out !== '{}') {
      this.writeAttr('x-globals', out);
    }
  }

  xArgs(args: ITypeArg[]): void {
    const value = srzArgs(args, this);
    if (value) {
      this.writeDirective('x-args', value);
    }
  }

  newline(): void {
    if (!this.compress) {
      this.html += '\n';
    }
  }

  newIndent(): void {
    if (!this.compress) {
      this.html += this.indent;
    }
  }
}

function isIdentTag(tag: string) {
  return tag !== 'html' && tag !== 'head' && tag !== 'body';
}
