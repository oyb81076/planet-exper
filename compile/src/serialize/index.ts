import {
  IHTMLFragDocument, IHTMLFullDocument, IChildNode, NodeTypes,
} from '../ast';
import Serializer from './Serializer';
import type { ISerializeConfig } from './faces';

export default function serialize(
  doc: IHTMLFragDocument | IHTMLFullDocument,
  config: ISerializeConfig = {},
): string {
  const { element } = doc;
  const ctx = new Serializer(config);
  if (doc.type === '#html-document') {
    ctx.doctype();
  }
  ctx.openTag(element.tagName);
  ctx.xKey(element.key);
  if (doc.type === '#html-fragment') {
    ctx.xArgs(doc.args);
  }
  ctx.xLinks(doc.links);
  ctx.xScripts(doc.scripts);
  ctx.xJsGlobals(doc.jsGlobals);
  ctx.attrs(element.attrs);
  ctx.directives(element.directives);
  writeNodes(ctx, element.nodes);
  ctx.closeTag();
  return ctx.html;
}

// 这里使用去递归的方式调用
function writeNodes(ctx: Serializer, nodes: IChildNode[]) {
  const undo = createUndo(nodes);
  let index = undo.length - 1;
  while (index !== -1) {
    const node = undo[index];
    if (node.opened) {
      ctx.closeTag();
      undo.pop();
      index -= 1;
    } else if (node.element.type === NodeTypes.TEXT) {
      ctx.text(node.element.content);
      undo.pop();
      index -= 1;
    } else {
      node.opened = true;
      const { element } = node;
      ctx.openTag(element.tagName);
      ctx.xKey(element.key);
      ctx.attrs(element.attrs);
      ctx.directives(element.directives);
      const nodesLength = element.nodes.length;
      if (!nodesLength) {
        ctx.closeTag();
        undo.pop();
        index -= 1;
      } else if (nodesLength === 1 && element.nodes[0].type === NodeTypes.TEXT) {
        ctx.inlineText(element.nodes[0].content);
        ctx.closeTag();
        undo.pop();
        index -= 1;
      } else {
        undo.push(...createUndo(element.nodes));
        index += element.nodes.length;
      }
    }
  }
}
function createUndo(nodes: IChildNode[]) {
  return nodes.slice(0).reverse().map((x) => ({ element: x, opened: false }));
}
