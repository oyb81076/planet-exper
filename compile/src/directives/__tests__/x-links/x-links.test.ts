import { readFileSync } from 'fs';
import { join } from 'path';
import { compileFull } from '../../../compile';
import serialize from '../../../serialize';

describe('single', () => {
  const SRC_HTML = readFileSync(join(__dirname, './x-links-single.html'), 'utf-8');
  it('all', () => {
    const doc = compileFull(SRC_HTML, { parseExpression: true });
    const loose = serialize(doc, { compress: false, quote: true });
    expect(loose).toMatchSnapshot();
    const compact = serialize(doc, {
      compress: false, quote: true, compactJS: true, compactJSON: true,
    });
    expect(compact).toMatchSnapshot();
  });
});

describe('multiple', () => {
  const SRC_HTML = readFileSync(join(__dirname, './x-links-multiple.html'), 'utf-8');
  it('all', () => {
    const doc = compileFull(SRC_HTML, { parseExpression: true });
    const loose = serialize(doc, { compress: false, quote: true });
    expect(loose).toMatchSnapshot();
    const compact = serialize(doc, {
      compress: false, quote: true, compactJS: true, compactJSON: true,
    });
    expect(compact).toMatchSnapshot();
  });
});
