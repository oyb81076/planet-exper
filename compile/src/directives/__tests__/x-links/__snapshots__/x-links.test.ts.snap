// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`multiple all 1`] = `
<!DOCTYPE html>
<html x-key="pgjdSF" x-links="[
  '5ee71f0159c61b62cdaaf7a3',
  {href:'http://www.baidu.com'},
  {media:'screen and (min-width: 600px) and (max-width: 800px)',file:'5ee71f0159c61b62cdaaf7a3'}
]">
<head x-key="LPxZ84"></head>
<body x-key="joS4MV"></body>
</html>
`;

exports[`multiple all 2`] = `
<!DOCTYPE html>
<html x-key="pgjdSF" x-links="['5ee71f0159c61b62cdaaf7a3',{href:'http://www.baidu.com'},{media:'screen and (min-width: 600px) and (max-width: 800px)',file:'5ee71f0159c61b62cdaaf7a3'}]">
<head x-key="LPxZ84"></head>
<body x-key="joS4MV"></body>
</html>
`;

exports[`single all 1`] = `
<!DOCTYPE html>
<html x-key="QMhuN1" x-links="5ee71f0159c61b62cdaaf7a3">
<head x-key="Sea6jl"></head>
<body x-key="pe0F3i"></body>
</html>
`;

exports[`single all 2`] = `
<!DOCTYPE html>
<html x-key="QMhuN1" x-links="5ee71f0159c61b62cdaaf7a3">
<head x-key="Sea6jl"></head>
<body x-key="pe0F3i"></body>
</html>
`;
